src/coreclr/vm/pgo.cpp

```diff
@@ -285,206 +322,516 @@ void PgoManager::ReadPgoData()
+    {
+        LayoutPgoInstrumentationSchema(prevSchema, &pSchema[iSchema]);
+        prevSchema = pSchema[iSchema];
+    }
 
-    unsigned index = 0;
-    unsigned methodsChecked = 0;
+    S_SIZE_T allocationSize = S_SIZE_T(sizeof(HeaderList)) + S_SIZE_T(pSchema[countSchemaItems - 1].Offset + InstrumentationKindToSize(pSchema[countSchemaItems - 1].InstrumentationKind));
+    
+    if (allocationSize.IsOverflow())
+    {
+        return E_NOTIMPL;
+    }
+    size_t unsafeAllocationSize = allocationSize.Value();
+    HeaderList* pHeaderList = NULL;
 
-    while (index < maxIndex)
+    if (pMD->IsDynamicMethod())
     {
-        // The first two "records" of each entry are actually header data
-        // to identify the method.
-        //
-        Header* const header = (Header*)&s_PgoData[index];
+        HeaderList *currentHeaderList = m_pgoHeaders;
+        if (currentHeaderList != NULL)
+        {
+            if (!ComparePgoSchemaEquals(currentHeaderList->header.GetData(), currentHeaderList->header.countsOffset, pSchema, countSchemaItems))
+            {
+                return E_NOTIMPL;
+            }
+            _ASSERTE(currentHeaderList->header.method == pMD);
+            *pInstrumentationData = currentHeaderList->header.GetData();
+            return S_OK;
+        }
 
-        // Sanity check that header data looks reasonable. If not, just
-        // fail the lookup.
-        //
-        if ((header->recordCount < MIN_RECORD_COUNT) || (header->recordCount > MAX_RECORD_COUNT))
+        pHeaderList = (HeaderList*)pMD->AsDynamicMethodDesc()->GetResolver()->GetJitMetaHeap()->New(unsafeAllocationSize);
+
+        memset(pHeaderList, 0, unsafeAllocationSize);
+        pHeaderList->header.Init(pMD, codehash, ilSize, offsetOfInstrumentationDataFromStartOfDataRegion);
+        *pInstrumentationData = pHeaderList->header.GetData();
+        if (!WriteInstrumentationSchema(pSchema, countSchemaItems, *pInstrumentationData, pHeaderList->header.countsOffset))
         {
-            break;
+            _ASSERTE(!"Unable to write schema");
+            return E_NOTIMPL;
         }
+        m_pgoHeaders = pHeaderList;
+        return S_OK;
+    }
+    else
+    {
+        LoaderAllocatorPgoManager *laPgoManagerThis = (LoaderAllocatorPgoManager *)this;
+        CrstHolder (&laPgoManagerThis->m_lock);
 
-        // See if the header info matches the current method.
-        //
-        if ((header->token == token) && (header->hash == hash) && (header->ilSize == ilSize))
+        HeaderList* existingData = laPgoManagerThis->m_pgoDataLookup.Lookup(pMD);
+        if (existingData != NULL)
         {
-            // Yep, found data.
-            //
-            *pBlockCounts = &s_PgoData[index + 2];
-            *pCount       = header->recordCount - 2;
-            *pNumRuns     = 1;
+            if (!ComparePgoSchemaEquals(existingData->header.GetData(), existingData->header.countsOffset, pSchema, countSchemaItems))
```

Comment:
Ditto here, do we need to make sure we're looking at the same IL version the jit is compiling?

Reply:
I wouldn't expect so. The pgo instrumentation data structures will be the same if the schema's match up, and the JIT already must tolerate data which isn't completely accurate.